File: demo-iys-rest-scraping.java/src/main/java/demo/reader/AbstractReader.java
package demo.reader;

import com.google.api.client.util.Lists;
import com.google.common.collect.Sets;
import demo.executor.Executor;
import demo.executor.ParallelExecutor;
import demo.file.TextFile;
import demo.http.HttpRequester;
import demo.js.SkillsParser;
import demo.json.ProgressFile;
import demo.json.TargetsListFile;
import demo.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

public abstract class AbstractReader implements Reader {

    private static final Logger logger = LoggerFactory.getLogger(AbstractReader.class);

    private static final int PARALLEL_EXECUTORS_NUMBER = 10;

    private final String url;
    private final String name;

    private Executor<SourcesList> executor;
    private SkillsParser skillsParser;
    private ProgressFile progressFile;
    private AtomicInteger progress = new AtomicInteger(0);
    private TargetsListFile targetsListFile;
    private TargetsList targetsList = new TargetsList();
    private TextFile dictionaryFile;
    private List<String> dictionary;

    public AbstractReader(String url, String name) throws Exception {
        this.url = url;
        this.name = name;

        this.executor = new ParallelExecutor<>(PARALLEL_EXECUTORS_NUMBER);
        this.skillsParser = new SkillsParser();
        this.dictionaryFile = new TextFile("reader/dictionary.txt");

        this.progressFile = new ProgressFile("progress/progress.json");
        if (this.progressFile.load(name) == null) {
            List<String> combinations = new TextFile("combination/combinations1.txt").loadAsList();
            combinations.addAll(new TextFile("combination/combinations2.txt").loadAsList());
            this.dictionaryFile.save(combinations);
        }
        this.progressFile.save(name, progress.get());

        this.targetsListFile = new TargetsListFile(name);
        this.targetsList = targetsListFile.load();

        this.dictionary = this.dictionaryFile.loadAsList();
    }

    @Override
    public void read() {
        final Set<Target> targets = Sets.newCopyOnWriteArraySet(targetsList.getTargets());

        while (progress.get() < dictionary.size()) {
            executor.execute(
                    () -> {
                        final String code = dictionary.get(progress.getAndIncrement());
                        logger.info("progress " + progress + "/" + dictionary.size() + String.format(" %.3f%% ", (double) 100 * progress.get() / dictionary.size()) + code);

                        if (code.contains("%")) {
                            return new SourcesList(Lists.newArrayList(), 0, code);
                        } else {
                            return readSources(String.format("&q=%s", code), code);
                        }
                    },
                    sourcesList -> {
                        updateTargets(targets, sourcesList);

                        final String code = sourcesList.getCode();
                        dictionary.remove(code);

                        if (!code.contains("&page=")) {
                            if (sourcesList.getTotal() > sourcesList.getSources().size()) {
                                final int pages = (sourcesList.getTotal() / sourcesList.getSources().size()) + 1;
                                for (int page = 2; page <= pages; page++) {
                                    dictionary.add(String.format("%s&page=%d", code, page));
                                }
                            }
                        }
                    }
            );
        }

        logger.info("Reader finished");
    }

    private SourcesList readSources(String suffix, String code) throws IOException {
        logger.debug("request: " + suffix);
        Response response = HttpRequester.request(url + suffix);
        logger.debug("response: " + code + " " + response.status + " " + response.total + " " + response.skills);
        return new SourcesList(skillsParser.parse(response.skills), response.total, code);
    }

    private synchronized void updateTargets(Set<Target> targets, SourcesList sourcesList) {
        int sizeBefore = targets.size();

        if (sourcesList.getSources().size() > 0) {
            logger.info("returned " + sourcesList.getSources().size() + "/" + sourcesList.getTotal());
        }

        for (Source source : sourcesList.getSources()) {
            targets.add(new Target(source.id.trim(), source.text.trim()));
        }

        int sizeAfter = targets.size();
        if (sizeAfter > sizeBefore) {
            logger.info("added " + (sizeAfter - sizeBefore));

            progressFile.save(name, progress.get());
            targetsList.updateTargets(targets);
            targetsListFile.save(targetsList);

            dictionaryFile.save(dictionary);
        }
    }
}

File: demo-iys-rest-scraping.kotlin/src/main/kotlin/demo/reader/AbstractReader.kt
package demo.reader

import com.google.api.client.util.Lists
import com.google.common.collect.Sets
import demo.executor.Executor
import demo.executor.ParallelExecutor
import demo.file.TextFile
import demo.http.HttpRequester
import demo.js.SkillsParser
import demo.json.ProgressFile
import demo.json.TargetsListFile
import demo.model.SourcesList
import demo.model.TargetsList
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.concurrent.Callable
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer

abstract class AbstractReader @Throws(Exception::class)
constructor(private val url: String, private val name: String) : Reader {

    private val executor: Executor<SourcesList>
    private val skillsParser: SkillsParser
    private val progressFile: ProgressFile
    private val progress = AtomicInteger(0)
    private val targetsListFile = TargetsListFile(name)
    private val targetsList = targetsListFile.load()
    private val dictionaryFile: TextFile
    private val dictionary: MutableList<String>

    init {
        this.executor = ParallelExecutor<SourcesList>(PARALLEL_EXECUTORS_NUMBER)
        this.skillsParser = SkillsParser()
        this.dictionaryFile = TextFile("reader/dictionary.txt")

        this.progressFile = ProgressFile("progress/progress.json")
        if (this.progressFile.load(name) == null) {
            val combinations = TextFile("combination/combinations1.txt").loadAsList()
            combinations.addAll(TextFile("combination/combinations2.txt").loadAsList())
            this.dictionaryFile.save(combinations)
        }
        this.progressFile.save(name, progress.get())

        this.dictionary = this.dictionaryFile.loadAsList()
    }

    override fun read() {
        val targets = Sets.newCopyOnWriteArraySet(targetsList.targets)

        while (progress.get() < dictionary.size) {
            executor.execute(
                    Callable<SourcesList> {
                        val code = dictionary[progress.andIncrement]
                        logger.info("progress " + progress + "/" + dictionary.size + String.format(" %.3f%% ", 100.toDouble() * progress.get() / dictionary.size) + code)

                        if (code.contains("%")) {
                            SourcesList(Lists.newArrayList(), 0, code)
                        } else {
                            readSources(String.format("&q=%s", code), code)
                        }
                    },
                    Consumer<SourcesList> { sourcesList ->
                        updateTargets(targets, sourcesList)

                        val code = sourcesList.code
                        dictionary.remove(code)

                        if (!code.contains("&page=")) {
                            if (sourcesList.total > sourcesList.sources.size) {
                                val pages = sourcesList.total / sourcesList.sources.size + 1
                                for (page in 2..pages) {
                                    dictionary.add(String.format("%s&page=%d", code, page))
                                }
                            }
                        }
                    }
            )
        }

        logger.info("Reader finished")
    }

    @Throws(IOException::class)
    private fun readSources(suffix: String, code: String): SourcesList {
        logger.debug("request: " + suffix)
        val response = HttpRequester.request(url + suffix)
        logger.debug("response: " + code + " " + response.status + " " + response.total + " " + response.skills)
        return SourcesList(skillsParser.parse(response.skills!!), response.total, code)
    }

    @Synchronized private fun updateTargets(targets: MutableSet<demo.model.Target>, sourcesList: SourcesList) {
        val sizeBefore = targets.size

        if (sourcesList.sources.size > 0) {
            logger.info("returned " + sourcesList.sources.size + "/" + sourcesList.total)
        }

        for (source in sourcesList.sources) {
            targets.add(demo.model.Target(source.id!!.trim(), source.text!!.trim()))
        }

        val sizeAfter = targets.size
        if (sizeAfter > sizeBefore) {
            logger.info("added " + (sizeAfter - sizeBefore))

            progressFile.save(name, progress.get())
            targetsList.updateTargets(targets)
            targetsListFile.save(targetsList)

            dictionaryFile.save(dictionary)
        }
    }

    companion object {

        private val logger = LoggerFactory.getLogger(AbstractReader::class.java)

        private val PARALLEL_EXECUTORS_NUMBER = 10
    }
}